Fix error handling in X509_chain_up_ref (attempt 2)

mattcaswell opened the pull request:
We were ignoring the return value from  X509_up_ref. This rewrites the function and handles errors from X509_up_ref correctly.

It also adds a test.

Fixes #8592

This is an alternative approach to that given in #8598.

paulidale commented on crypto/x509/x509_cmp.c:
```diff
@@ -440,19 +440,28 @@ int X509_CRL_check_suiteb(X509_CRL *crl, EVP_PKEY *pk, unsigned long flags)
-    }
-    return ret;
+    /* We upref each element rather than actually copying */
+    return sk_X509_deep_copy(chain, x509upref, X509_free);
```
Would the const casting be better done to the function pointer here?

mattcaswell replied:
Yes. Actually that's a great idea. Done.

kroeckx commented on crypto/x509/x509_cmp.c:
```diff
@@ -440,19 +440,23 @@ int X509_CRL_check_suiteb(X509_CRL *crl, EVP_PKEY *pk, unsigned long flags)
-    }
-    return ret;
+    /* We upref each element rather than actually copying */
+    return sk_X509_deep_copy(chain, (X509 *(*)(const X509 *))x509upref,
```
I'm not a big fan of pretending that we're not modifying the original, and have to wonder how the compiler might optimize it.

mlindner commented on crypto/x509/x509_cmp.c:
```diff
@@ -440,19 +440,23 @@ int X509_CRL_check_suiteb(X509_CRL *crl, EVP_PKEY *pk, unsigned long flags)
-    }
-    return ret;
+    /* We upref each element rather than actually copying */
+    return sk_X509_deep_copy(chain, (X509 *(*)(const X509 *))x509upref,
```
I'm also not a big fan of abusing the deep_copy function this way. It confuses the reader when they're reading the code thinking its doing a deep copy and not just a shallow copy like the original code implied.

mlindner replied:
Arguably the sk_X509_dup function is misnamed as a shallow-duplicated stack should involve bumping the reference counts of all elements automatically.

levitte replied:
One might argue that `sk_*_deep_copy` is misnamed, and indicates an intended use rather than the actual functionality.  Basically, the "depth" of the copy is governed by the supplied copyfunc rather than by `sk_*_deep_copy` itself.

The right thing to do would probably be to rename `sk_*_deep_copy` to just `sk_*_copy`, and provide the former name as a (deprecated?) alias.

mlindner replied:
That would make sense, but this needs to be backported to 1.1.1.

levitte replied:
That's fine.  I think we can live with this "misuse" of `sk_*_deep_copy`...  but perhaps a comment that remarks on this would be suitable.

paulidale replied:
Agreed, a comment would be beneficial.

mattcaswell commented:
I expanded out the comment in the code which hopefully addresses most of the concerns. I'm not sure whether @kroeckx's comment amounts to an actual objection to this PR. I don't think it does, but I'll leave this PR open over the weekend at least, just in case.

@paulidale - please reconfirm?

kroeckx commented:
I would prefer a new function that didn't take a const argument in
the copy function.

mattcaswell commented:
> I would prefer a new function that didn't take a const argument in the copy function.

Do you mean a new function like sk_\*\_deep_copy? e.g. perhaps just sk_\*\_copy? If so that could probably be done for 3.0 but can it be justified for backport to 1.1.1?

kroeckx commented:
> Do you mean a new function like sk_\*\_deep_copy? e.g. perhaps just sk_\*\_copy?

Yes.

> If so that could probably be done for 3.0 but can it be justified for backport to 1.1.1?

You can just do what that function does inside the
X509_chain_up_ref() for 1.1.1.

FdaSilvaYY commented on crypto/x509/x509_cmp.c:
```diff
@@ -440,19 +440,29 @@ int X509_CRL_check_suiteb(X509_CRL *crl, EVP_PKEY *pk, unsigned long flags)
-    return ret;
+    /*
+     * We upref each element rather than actually copying. We're "misusing"
+     * sk_X509_deep_copy a little here since we're not actually dong a "deep"
```
actually **dong** ?

mattcaswell commented:
Closing this. It was eventually fixed by #9614 anyway.
